Added support for underscored(All):true

leroydev opened the pull request:
I think this would fix #18, I've added tests for it but I'm not sure if I've covered everything needed in order to guarantee that this change covers all field names.

leroydev commented:
Oh.. I just noticed Sequelize v2 doesn't have the underscoring function I've used.. I'll try and find a workaround for that.

overlookmotel commented:
@leroydev Thanks for this PR. I'll have a look in next few days.

Concerning the `Utils._.underscoredIf` / `Utils.underscoredIf` problem, there's quite a lot of patches in the module where API changed between Sequelize versions. Have a look in [lib/patches.js](https://github.com/overlookmotel/sequelize-hierarchy/blob/master/lib/patches.js). What I'd suggest is add a method `patches.underscoredIf()` which calls `Utils._.underscoredIf` or `Utils.underscoredIf` depending on Sequelize version - and then call `patches.underscoredIf()` from within the main code.

overlookmotel commented:
I guess Sequelize v2 support could be dropped at some point, but I think not until Sequelize v4 final drops.

overlookmotel commented:
One other thing: I released v0.8.0 last night which included a lot of changes to support Sequelize v4.0 - including a lot of changes in `lib/patches.js`. Can you please rebase your branch on to latest master before you make more changes?

leroydev commented:
@overlookmotel Rebase should have succeeded. I also added the patch for `underscoredIf`. Let me know if there's something missing or if something can be improved. :)

overlookmotel commented:
Thanks very much for your work on this.

I hope you don't mind if I give you some fairly detailed feedback.

Here are my thoughts:

* `underscored` option should only affect column names not table names, so I think `options.throughTable` should be unaffected.
* `model.options.underscored` option is inherited from `sequelize.options.underscored`. I think this happens automatically but could you please check this works? No need to write a test for it, but could you please check and confirm? i.e. that if you set `sequelize.options.underscored = true` and then leave `model.underscored` undefined, the underscoring still happens.
* There is also a option on `sequelize` (rather than the model) called `underscoredAll` (see the docs at http://sequelize.readthedocs.io/en/latest/api/sequelize/) which affects table names, so I guess `options.throughTable` should be underscored or not depending on *that* option rather than by `model.options.underscored`.
* `options.through` is the model name not a table name and so should not be affected by either `underscored` or `underscoredAll` options.
* The code as you have it now alters the field names even if the user has specified a value for them with e.g. `option.levelFieldName`. I don't think this is good - if the user specifies what they want to field name to be, we should respect that. So the `underscoreIf` calls should happen on the defaults only in my opinion. If the option is specified leave it alone; if it's unspecified and uses the default, then `options.underscored` should be used to set the value. (does that make sense?)
* I think you can make lines 53-57 into a one-liner `var genThroughTableName = patches.underscoredIf(Utils.pluralize(this.name) + (options.camelThrough ? Utils.uppercaseFirst(options.ancestorsAs) : options.ancestorsAs), this.options.underscored);`, or (even better) combine it with the next statement and do away with using the temporary `getThroughTableName` variable at all.

Does that all make sense? I'm sorry if I'm being overly precise but I don't use the `underscored` options myself so i's not a priority for me to implement this. But what I *do* really care about is, if it is implemented, it is done perfectly otherwise someone else will end up posting an issue that it doesn't work right and it'll be more work to fix it. Anyway, getting things exactly right is always good!

The reason I never got around to solving issue #18 is that I found it more complicated than I'd expected with the `underscored` and `underscoredAll` options and I worried about getting it wrong, so I left it. I'm very glad you're getting on the case though!

leroydev commented:
I set `underscored: true` in `sequelizeOptions` in support.js and removed `underscored: true` from my tests model definition, which resulted in `sequelize.options.underscored` being true but `this.folder.options.underscored` being false. Should I fix this in some way?

I think `through` should remain untouched by `underscored` too, is that correct?

The one-liner you proposed doesn't result in underscoring. I introduced the `genThroughTableName` variable on purpose so it wouldn't turn into a big, hard to read statement. I don't think two ternary operators in one line of code would improve readability, what are your thoughts on this?

User supplied options should indeed have priority over `underscored: true`, I'll fix that and write tests for it. 👍 

Would it be okay for this pull request to handle `underscored` but let `underscoredAll` be handled in another pull request? Or would you rather have both of them handled in this pull request?

leroydev commented:
The failing tests seem unrelated to my changes, I'm not 100% sure though, what do you think?

overlookmotel commented on lib/patches.js:
```diff
@@ -18,6 +18,13 @@ module.exports = function(Sequelize) {
+         */
+        underscoredIf: {
+          '^2.0.0': Sequelize.Utils._.underscoredIf,
+          '* || >=4.0.0-1': Sequelize.Utils.underscoredIf
```
`'* || >=4.0.0-1'` should be `'>=3.0.0 || ^4.0.0-0'` to match the rest (and so not to match v1.x.x.

overlookmotel replied:
NB this changed for the other patches in v0.8.0

overlookmotel commented:
Great stuff! This looks great now.

To answer your questions/notes:

> I set `underscored: true` in `sequelizeOptions` in support.js and removed `underscored: true` from my tests model definition, which resulted in `sequelize.options.underscored` being true but `this.folder.options.underscored` being false. Should I fix this in some way?

Ah. I guess the code needs to check `sequelize.options.underscored` if `model.options.underscored` is undefined. I'd suggest adding a line at the top:

```js
var underscored = this.options.underscored === undefined ? this.sequelize.options.underscored : this.options.underscored;
```

and then use `underscored` var after that.

No need to write tests for this (and it'd be hard to) but please can you check it works?

> Would it be okay for this pull request to introduce `underscored` support but introduce `underscoredAll` support in another pull request? Or would you rather have both of them handled in this pull request?

Well it's your PR! Sure it'd be great if both options were supported but it's totally up to you. If you don't have time to do `underscoredAll` too, then feel free to leave it for another day/someone else.

> The one-liner you proposed doesn't result in underscoring. I introduced the `genThroughTableName` variable on purpose so it wouldn't turn into a big, hard to read statement. I don't think two ternary operators in one line of code would improve readability, what are your thoughts on this?

Gah! Of course you're right. My bad. Two ternary operators: well, yeah, it's a matter of taste. I sometimes do it, but then split the expression over a several lines so it's readable. Anyway, my suggested expression doesn't work anyway, so this is a moot point.

> Does this option need to be documented in README.md? I could imagine documenting it being redundant if we handle the option the same way Sequelize does it.

I don't think any README mention needed. As you say, it's just what you'd expect from standard sequelize behaviour. Please add a changelog entry though mentioning "(fixes #18)"

NB I released v1.0.0 this morning so I'm afraid you'll need to rebase again. There shouldn't be any conflicts as long as you do the rebase before adding to changelog.

Please can you also squash all the commits into 1, so it's ready for merging?

overlookmotel commented:
Travis test fails... really weird. Failing on MySQL only on all versions of node and sequelize. I think it's unrelated to these changes - probably Travis having a brain fart. I'll re-run the failed jobs tomorrow and see if they pass this time.

leroydev commented on test/all.test.js:
```diff
@@ -86,9 +86,34 @@ function tests() {
       expect(this.folder.associations.ancestors.through.model.attributes).to.have.property('testThroughForeignKey');
       expect(this.folder.associations.ancestors.foreignKey).to.equal('testThroughKey');
       expect(this.folder.associations.children.foreignKey).to.equal('testForeignKey');
+			expect(this.folder.associations.ancestors.through.model.tableName).to.equal('foldersancestors');
```
I'm not sure why this line is indented improperly, it showed up fine for me in my editor (Atom)

leroydev commented:
Did the squash and commit reword work as it's supposed to? I didn't do a squash before so I'm not sure if I did it correctly.

overlookmotel commented on changelog.md:
```diff
@@ -334,3 +334,7 @@ Breaking changes:
 * Refactor `lib/errors`
 * Increase tests timeout to 30 seconds
+
+## 1.0.1
```
Could you possible change this line to `## Next`? I add the version tag to changelog when I release. Anyway, I think this qualifies for a `1.1.0` release!

leroydev replied:
Will do.

overlookmotel commented on lib/modelExtends.js:
```diff
@@ -27,13 +27,17 @@ module.exports = function(Sequelize) {
 			var globalOptions = sequelize.options.define || {};
 
+			//If underscored(All) is set to true as global Sequelize option we let it override the model option
+			var underscored = this.options.underscored || this.sequelize.options.underscored;
```
This should be `var underscored = this.options.underscored === undefined ? this.options.underscored : this.sequelize.options.underscored;`

The current line will defer to `sequelize.options.underscore` if `model.options.underscored == false` which isn't right. `== false` means definitively no underscoring.

overlookmotel commented on lib/modelExtends.js:
```diff
@@ -44,14 +48,20 @@ module.exports = function(Sequelize) {
 			});
 
+			var genThroughTableName;
+			if (underscoredAll)
```
Would you mind putting curly brackets `{}` around the if and else clauses please? (in line with rest of the code style)

leroydev replied:
Ok, will do. :)

overlookmotel replied:
Also, should this use `this.sequelize.options.define.underscoredAll` rather than `this.options.underscoredAll`? The latter is an option set on the main model. The through table is a *different* model so I guess should inherit from the global options. NB this is aliased to local var `globalOptions` further up this function's body.

leroydev replied:
I'll simply name it `throughTable`.

That sounds reasonable, I think it'd need to be documented in that case though. 

If we were to do this, `this.sequelize.options.define` is empty by default. Should I let the `throughTable` name be underscored if `this.sequelize.options.define` has property `underscoredAll` **and** the properties value is `true`?

leroydev replied:
No haha, that's not what I meant. There's already a `throughTable` option which (even with this PR) behaves as you described.

You proposed to use `this.sequelize.options.define.underscoredAll` instead of `this.options.underscoredAll` for applying underscoring to the `throughTable`. To that, I replied:

> That sounds reasonable, I think it'd need to be documented in that case though.
> 
> If we were to do this, this.sequelize.options.define is empty by default. Should I let the throughTable name be underscored if this.sequelize.options.define has property underscoredAll and the properties value is true?

overlookmotel replied:
Ah I see. I completely forgot I implemented a `throughTable` option!

Yes, exactly how you said with `underscoredAll` option. `underscoredAll = !!(this.sequelize.options.define || {} ).underscoredAll`

leroydev commented:
For your interest, [this issue](It might be good to be aware of [this issue](https://github.com/sequelize/sequelize/issues/6619) is probably going to affect this pull request in the future if it gets fixed.

overlookmotel commented on test/all.test.js:
```diff
@@ -51,6 +51,69 @@ describe(Support.getTestDialectTeaser('Tests'), function () {
 
 function tests() {
+	describe('underscored:true', function() {
+		/*jshint camelcase: false */
```
Missing space before `jshint`

overlookmotel commented on test/all.test.js:
```diff
@@ -51,6 +51,69 @@ describe(Support.getTestDialectTeaser('Tests'), function () {
+
+			expect(this.folder.attributes).to.have.property('testFieldName');
+			expect(this.folder.attributes).to.have.property('testForeignKey');
+			expect(this.folder.associations.ancestors.through.model.attributes).to.have.property('testThroughForeignKey');
```
Lines 85 and 86 are the same. Is one of them mean to test fro `testThroughKey`?

leroydev replied:
No, I copied
```
expect(this.folder.associations.ancestors.through.model.attributes).to.have.property('folder_id');
expect(this.folder.associations.ancestors.through.model.attributes).to.have.property('ancestor_id');
```
And changed those property names to `testThroughForeignKey` for the "doesn't override user supplied names" test. Changing those property names made made one of the assertions redundant. I'll remove one of them.

overlookmotel commented:
Thanks for seeing this through. It's been a surprisingly long process for such a small point. Now I remember why I got annoyed and gave up trying to fix this previously. Sorry I've been picky.

If you agree to the line comments I've made, and are happy to make those changes, I think it's good to merge.

overlookmotel commented:
The squash worked great by the way. Thanks for that.

overlookmotel commented:
I'm afraid you'll have to rebase again because I just merged your other PR. The only conflict will be in `changelog.md`

leroydev commented:
What do you think about what I said [here](https://github.com/overlookmotel/sequelize-hierarchy/pull/80#discussion_r80233046)?

overlookmotel commented:
@leroydev Sorry I hadn't notices that comment. Sorry I was slow to reply. Have replied [here](https://github.com/overlookmotel/sequelize-hierarchy/pull/80#discussion_r81311035)

overlookmotel commented:
Hi @leroydev.

How are you getting on with this? Looks to me like it's ready to merge, but I don't know if you had anything further you wanted to do on it?

leroydev commented:
@overlookmotel I wasn't sure if documentation needed to be added to the README, except for that I think it's done. :)

overlookmotel commented:
Don't worry I'll update the docs. Will merge it shortly.

leroydev commented:
@overlookmotel any updates on this? :)

overlookmotel commented:
Finally merged and published as v1.2.0.

I made some small tweaks, mostly code style, but didn't change the logic of your code.

Really sorry @leroydev that it's taken me so long. And thanks loads for the sterling work in getting this working.
